controller/replicator.go

```diff
@@ -312,6 +312,98 @@ func (r *replicator) Snapshot(name string, userCreated bool, created string) err
 	return nil
 }
 
+func (r *replicator) GetLatestSnapshot() (string, error) {
+	retErrorLock := sync.Mutex{}
+	retError := &BackendError{
+		Errors: map[string]error{},
+	}
+	replicaChains := make(map[string][]string)
+	wg := sync.WaitGroup{}
+
+	for addr, backend := range r.backends {
+		if backend.mode == types.RW {
+			wg.Add(1)
+			go func(address string, backend types.Backend) {
+				if chain, err := backend.GetReplicaChain(); err != nil {
+					logrus.Infof("failed getting replica chain from %s with err %v", address, err)
+					retErrorLock.Lock()
+					retError.Errors[address] = err
+					retErrorLock.Unlock()
+				} else {
+					retErrorLock.Lock()
+					replicaChains[address] = chain
+					retErrorLock.Unlock()
+				}
+				wg.Done()
+			}(addr, backend.backend)
+		} else {
+			logrus.Infof("not getting chain from %s as replica in %v mode", addr, backend.mode)
+			break
+		}
+	}
+	wg.Wait()
+	if len(retError.Errors) != 0 {
+		return "", retError
+	}
+	if len(replicaChains) != len(r.backends) {
+		return "", fmt.Errorf("Not able to fetch chain from all replicas")
+	}
+	var snapName string
+	for _, chain := range replicaChains {
+		if len(chain) <= 1 {
+			return "", fmt.Errorf("No snapshot present in one of the replicas: %v", replicaChains)
+		}
+		if snapName == "" {
+			snapName = chain[1]
+			continue
+		}
+		if snapName != chain[1] {
+			return "", fmt.Errorf("Latest Snapshot differs at replicas %v", replicaChains)
+		}
+	}
+	return snapName, nil
```

Comment:
Though checkpoints are same across replicas, it is possible that chain is not same.
Need to check if chain is matching across replicas before returning the check pointed snapshot.
Please clarify if i'm missing something and put a comment why this is required.

Reply:
This function is called when all the replicas are in rw, in which case the latest snapsohts in all the replicas should be same. And chain above the checkpoint will also be same, since all are in sync. Chain below the checkpoint might be different, but we are not concerned about that.